Flexible API for defining connection configuration

bilderbuchi opened the pull request:
Flexible API for defining connection configuration as per comment https://github.com/pymeasure/pymeasure/pull/536#discussion_r763372161

Connection configuration is now passed fully as kwargs to Instrument.init, and sorted out within VISAAdapter.
Rebased from an earlier iteration, will be completed and documented before review.

bilderbuchi commented on pymeasure/adapters/visa.py:
```diff
@@ -71,34 +76,35 @@ def has_supported_version():
+        :param visa_library: Optional pyvisa visa_library specifier for creating the resource manager.
         """
         # Get VISA interface type associated with resource_name
-        if isinstance(resource_name, int):
-            resource_name = "GPIB0::%d::INSTR" % resource_name
-        elif isinstance(resource_name, str):
-            pass
-        else:
-            # User defined adapter, do not touch anything
-            return
```
this can be removed because it duplicates what already happens in VISAAdapter right before this method is called.

msmttchr reviewed (commented):
@bilderbuchi, I am sending few comments on the proposed changes. Sharing my gut feeling: we are back at the `safeKeyword` approach.

bilderbuchi commented on pymeasure/adapters/visa.py:
```diff
@@ -71,34 +76,35 @@ def has_supported_version():
+            if key in pyvisa.constants.InterfaceType.__members__:
+                if getattr(pyvisa.constants.InterfaceType, key) is if_type:
+                    # For the expected interface, dump contents into received_kwargs first
+                    received_kwargs.update(received_kwargs[key])
```
Just a quick remark, will review in detail later: can you propose (e.g. as a comment here) test cases for any features you are missing? That would help finding a good resolution.

bilderbuchi replied:
Nevermind, I'm on it.

bilderbuchi replied:
> In addition to allow the test test_serial_custom_baud_rate_is_set to pass a specific parameter needs to be added in the instrument interface.

I'm sorry, I don't understand, can you unpack that a bit?

bilderbuchi replied:
I think I resolved this now. Do you agree?

msmttchr replied:
Yes I agree

bilderbuchi commented on pymeasure/adapters/visa.py:
```diff
@@ -71,34 +76,35 @@ def has_supported_version():
-    def setdefaults(resource_name, interface_type, user_kwargs, **kwargs):
-        """ Change user_kwargs to add defaults specified by user for the selected interface type
+    def setdefaults(resource_name, received_kwargs, resource_mgr=None, visa_library=None):
+        """ Remove invalid entries from the received keyword arguments.
```
Good point, I already have a solution in my head.

bilderbuchi replied:
I think I resolved this now. Do you agree?

bilderbuchi commented on pymeasure/adapters/visa.py:
```diff
@@ -57,6 +57,11 @@ def __init__(self, resource_name, visa_library='', preprocess_reply=None, **kwar
             resource_name = "GPIB0::%d::INSTR" % resource_name
         self.resource_name = resource_name
         self.manager = pyvisa.ResourceManager(visa_library)
+        # We can make that one non-static and use self.resource_name and self.manager,
+        # then it would get smaller.
+        # A better name would be "clean_arguments"
```
@msmttchr what do you prefer? Mind if I make this a class method, maybe _private to indicate internal use only, or do you see this being useful outside the class somewhere?

msmttchr replied:
I think that, according to your approach, this method doesn't make sense anymore, even as internal method. I would remove it altogether and replace it with inline code where is needed.

msmttchr commented:
Sorry for being cryptic. I meant that to have an overridable default for
baud_rate, you were forced to add a specific parameter with default value
and then use this parameter as argument of the baud_rate parameter in the
instrument definition.

In other words, the customer passes a parameter which have a default (this
is just python), this parameter is passed to  a new parameter called aslr
via dictionary and finally it is set via the setdefault call. In my
opinion, this is a bit clunky and not straightforward, but of course this
is a matter of taste.

Il Mer 8 Dic 2021, 20:59 Christoph Buchner ***@***.***> ha
scritto:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pymeasure/adapters/visa.py
> <https://github.com/pymeasure/pymeasure/pull/539#discussion_r765203542>:
>
> > -
> -        if_type = pyvisa.ResourceManager(**visa_kwargs).resource_info(resource_name).interface_type
> +        if resource_mgr is None:
> +            visa_kwargs = {}
> +            if visa_library:
> +                visa_kwargs['visa_library'] = visa_library
> +            resource_mgr = pyvisa.ResourceManager(**visa_kwargs)
> +
> +        if_type = resource_mgr.resource_info(resource_name).interface_type
> +
> +        for key in list(received_kwargs):
> +            # Remove all interface-specific kwargs:
> +            if key in pyvisa.constants.InterfaceType.__members__:
> +                if getattr(pyvisa.constants.InterfaceType, key) is if_type:
> +                    # For the expected interface, dump contents into received_kwargs first
> +                    received_kwargs.update(received_kwargs[key])
>
> In addition to allow the test test_serial_custom_baud_rate_is_set to pass
> a specific parameter needs to be added in the instrument interface.
>
> I'm sorry, I don't understand, can you unpack that a bit?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymeasure/pymeasure/pull/539#discussion_r765203542>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJAXVWXIQTA4S5CFQ3Y62DDUP62LDANCNFSM5JTQSXXQ>
> .
>

bilderbuchi commented:
> Sorry for being cryptic. I meant that to have an overridable default for
baud_rate, you were forced to add a specific parameter with default value
and then use this parameter as argument of the baud_rate parameter in the
instrument definition.

Ah thanks. I didn't notice that adding a specific parameter with default value is necessary. In any case, now it is not necessary anymore. :-)

bilderbuchi commented:
@CasperSchippers @LongnoseRob please review this implementation a way to set/override connection defaults.

We defined tests that nail down the desired behaviour for users and instrument implementors.
I added a 7-line (without comments/docstrings) portion to VISAAdapter
I added and updated docs both from the user side ("how do I tweak connection settings for an existing instrument") and the implementor side ("how do I set up an instrument with proper connection defaults, both for single- and multiinterface devices).
Finally, I converted any instruments that I could find to the new best practice.

I will be happy to get any feedback on this. Did we miss a use case? Is the documentation clear? Can the docs be improved? Can the _implementation_ be improved?

LongnoseRob commented:
> @CasperSchippers @LongnoseRob please review this implementation a way to set/override connection defaults.
> 
> We defined tests that nail down the desired behaviour for users and instrument implementors. I added a 7-line (without comments/docstrings) portion to VISAAdapter I added and updated docs both from the user side ("how do I tweak connection settings for an existing instrument") and the implementor side ("how do I set up an instrument with proper connection defaults, both for single- and multiinterface devices). Finally, I converted any instruments that I could find to the new best practice.
> 
> I will be happy to get any feedback on this. Did we miss a use case? Is the documentation clear? Can the docs be improved? Can the _implementation_ be improved?

From just reading the changes it looks fine for me.
For testing it with the HP3478A: I might be able to do that later today or in the next days.

BTW: for the timeout on GPIB with pyvisa-py as backend (and maybe other backends aswell), the timeout cannot be set to arbitrary values. (Something that I learnt yesterday)

bilderbuchi commented:
> BTW: for the timeout on GPIB with pyvisa-py as backend (and maybe other backends aswell), the timeout cannot be set to arbitrary values. (Something that I learnt yesterday)

https://github.com/pyvisa/pyvisa-py/blob/40d439627aa6fd57f36093073c1e661f108cab17/pyvisa_py/gpib.py#L140

hm, this seems to be a limitation of a linux gpib library, and seems to be restricted to pyvisa-py and gpib, only. I can't find any indication otherwise in the pyvisa or pyvisa-py codebase.

LongnoseRob commented:
> > BTW: for the timeout on GPIB with pyvisa-py as backend (and maybe other backends aswell), the timeout cannot be set to arbitrary values. (Something that I learnt yesterday)
> 
> hm, this seems to be a [limitation of a linux gpib library](https://github.com/pyvisa/pyvisa-py/blob/40d439627aa6fd57f36093073c1e661f108cab17/pyvisa_py/gpib.py#L140), and seems to be restricted to pyvisa-py and gpib, only. I can't find any indication otherwise in the pyvisa or pyvisa-py codebase.

Yes, that is the same source I was looking at...
### Back to the topic:
I tested the  changes with [my verification file](https://github.com/LongnoseRob/misc/blob/3d26061585a6e8024044f873bbe4e1d3bdae26d3/3478_verifications.py) with shorted inputs (2 wire method)

output of the run  is [here](https://gist.github.com/LongnoseRob/d45dc1e9fa1b6f6e74cd7f6ccf7a77d2)
_(sorry it is a not very elaborate reporting test file)_

--> looks good to me and **"go"** for the merge from me

CasperSchippers reviewed (approved):
Nice job (to all, @bilderbuchi, @msmttchr, @LongnoseRob), and sorry for not contributing so far to this discussion.

This implementation looks pretty good to me; I like how it enables you to simply use keyword arguments to fully define the default communication parameters for all possible communication protocols, and similarly only need keyword arguments to overrule the defaults.

I didn't spot any issues or improvements. So for me it seems good to go.

PS I tested it with the Oxford instrument ITC503 (the only device that has been changed in this PR that I have access to) and it works perfectly fine as far as I can tell.

bilderbuchi commented on pymeasure/instruments/instrument.py:
```diff
@@ -35,14 +35,35 @@
 
-    :param adapter: An :class:`Adapter<pymeasure.adapters.Adapter>` object
-    :param name: A string name
+    It makes use of an array of :py:class:`~pymeasure.adapters.Adapter` classes for communication
```
Thanks, done!

bilderbuchi commented on pymeasure/adapters/visa.py:
```diff
@@ -57,6 +81,20 @@ def __init__(self, resource_name, visa_library='', preprocess_reply=None, **kwar
+
+        # Clean up kwargs considering the interface type matching resource_name
+        if_type = self.manager.resource_info(self.resource_name).interface_type
+        for key in list(kwargs):
```
Indeed, I agree.

bilderbuchi commented on pymeasure/adapters/visa.py:
```diff
@@ -57,6 +81,20 @@ def __init__(self, resource_name, visa_library='', preprocess_reply=None, **kwar
+                    # kwargs passed to Instrument.__init__()
+                    for k, v in kwargs[key].items():
+                        kwargs.setdefault(k, v)
+                del kwargs[key]
```
Yes it does, I added a test for this nonetheless

bilderbuchi commented:
Thanks for the review, I'll merge this!
